### DISCUSSION TITLE
RAM & Cpu Usage Optimisation

### DISCUSSION THREAD

**Discussion by harshakhmk** [milvus-io/milvus#30433] [Q&A and General discussion]
Hi, I am exploring milvus with docker and I have seen option to restrict Cpu and memory, but I was more curious to know if there are any other options to reduce the cpu and memory required to run milvus server 
Also while building from source is there any option to build only required modules without building all, as I have to build on lower end machines in arm Architecture

**Comment by yhmo** [comment]
We don't have any experience in deploying milvus on lower-end machines.
Milvus standalone is a docker container, if you assign 1 cpu core to the container, it will only use the 1 cpu core to work.
It requires data be loaded into memory for search, different indexes require different memory size.
Raw data size is calculated by row_count * dimension * 4bytes. (If you have long-length strings stored in varchar field, it will require extra memory size)
For FLAT/IVF_FLAT/HNSW, index size is almost equal to raw data size.
For IVF_SQ8/IVF_PQ, index size is about 25% ~ 30% of the raw data size.
For DISKANN, index size is about 1/6 ~ 1/4 of the raw data size.

Choose a proper index to save memory and match the recall rate you want.

I don't thing there are modules that can be ignored to build.

**Reply by harshakhmk** [reply]
Basically I want to make sure milvus server along with etcd and container servers all to run in less than 100 MB, during idle time 
Is it possible to achieve the same apart from changing cpu, memory parameters 

The indexes are more on doing optimisation on code side, but I am looking for options to do it while compiling/running the server 

Any specific core DB code level optimisation techniques

**Reply by yhmo** [reply]
milvus server and etcd are running in two different containers, you should configure them respectively.
You can limit milvus to 80MB, limit etcd to 20MB. I don't know whether the etcd can run with 20MB RAM. Milvus can start with 80MB RAM, but only a few data can be loaded into 80MB RAM. You can try, we can't ensure it works.

**Reply by harshakhmk** [reply]
Any other optimisation techniques you suggest specific to core DB level optimisation

**Reply by yhmo** [reply]
No, I don't have. Milvus is not designed to work under extremely limited resources.
With 100MB RAM, don't insert too many entities every time to avoid OOM. It could not process too many concurrent requests. Don't create too many collections/partitions.

**Reply by harshakhmk** [reply]
So you confirm that currently to limit cpu, RAM usage, only option is to update the YAML file by giving specified amount of cpu& memory variables
Kindly let us know if any other options are added in future 
Thanks

**Reply by yhmo** [reply]
Change this configuration to a small value in the milvus.yaml:
```
dataCoord:
  segment:
    maxSize: 16 # Maximum size of a segment in MB
```

I suggest changing it to 16 since the RAM is limited to no more than 100MB.

**Reply by harshakhmk** [reply]
Ok, any build flags or runtime flags are there to achieve the optimisation part in terms of CPU & RAM and any other optimisation

**Reply by yhmo** [reply]
No such flags

**Reply by harshakhmk** [reply]
Ok, do you have configuration file settings for getting better performance and optimisation that is ideal for Low computing machine

**Reply by yhmo** [reply]
No, I don't have. In fact, we didn't test milvus on low-end machines.

**Reply by harshakhmk** [reply]
I have applied the suggested change to milvus.yaml file, I have observed increased insert and search time
Any reasons

**Reply by yhmo** [reply]
If your data is not static, continually insert will exhaust the RAM eventually, and index tasks and compaction tasks also occupy CPU resources so that search performance is poor.

**Reply by harshakhmk** [reply]
I have tried the suggested change, no change in optimisation of CPU & RAM checker from docker stats
